In [1]:
from d3rlpy.datasets import get_cartpole

In [20]:
dataset, env = get_cartpole()

In [138]:
dataset.size()

1582

In [89]:
len(dataset.rewards)

99866

In [90]:
len(dataset.episodes)

1582

In [91]:
len(dataset.actions)

99866

In [139]:
from sklearn.model_selection import train_test_split

train_episodes, test_episodes = train_test_split(dataset, test_size=0.2)

In [140]:
from d3rlpy.algos import DQN

# if you don't use GPU, set use_gpu=False instead.
dqn = DQN(use_gpu=False)

# initialize neural networks with the given observation shape and action size.
# this is not necessary when you directly call fit or fit_online method.
dqn.build_with_dataset(dataset)

In [141]:
from d3rlpy.metrics.scorer import td_error_scorer
from d3rlpy.metrics.scorer import average_value_estimation_scorer

# calculate metrics with test dataset
td_error = td_error_scorer(dqn, test_episodes)

from d3rlpy.metrics.scorer import evaluate_on_environment

# set environment in scorer function, returns mean return per episode
evaluate_scorer = evaluate_on_environment(env)

# evaluate algorithm on the environment
rewards = evaluate_scorer(dqn)

In [142]:
env

<TimeLimit<CartPoleEnv<CartPole-v0>>>

In [143]:
# Use the output of learning, this is the mean reward and the most useful metric to measure learning
dqn.fit(train_episodes,
        eval_episodes=test_episodes,
        n_epochs=10,
        scorers={
            'td_error': td_error_scorer,
            'advantage': discounted_sum_of_advantage_scorer,
            'value_scale': average_value_estimation_scorer,
            'environment': evaluate_scorer
        })

2023-03-13 19:01:41 [debug    ] RoundIterator is selected.
2023-03-13 19:01:41 [info     ] Directory is created at d3rlpy_logs/DQN_20230313190141
2023-03-13 19:01:41 [warning  ] Skip building models since they're already built.
2023-03-13 19:01:41 [info     ] Parameters are saved to d3rlpy_logs/DQN_20230313190141/params.json params={'action_scaler': None, 'batch_size': 32, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 6.25e-05, 'n_critics': 1, 'n_frames': 1, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': None, 'target_update_interval': 8000, 'use_gpu': None, 'algorithm': 'DQN', 'observation_shape': (4,), 'action_size': 2}


Epoch 1/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:01:51 [info     ] DQN_20230313190141: epoch=1 step=2500 epoch=1 metrics={'time_sample_batch': 0.00012184162139892578, 'time_algorithm_update': 0.003392033576965332, 'loss': 0.012645094241556217, 'time_step': 0.00361170015335083, 'td_error': 0.9881827529171064, 'value_scale': 1.0015601515105368, 'environment': 9.4} step=2500
2023-03-13 19:01:51 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_2500.pt


Epoch 2/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:02:00 [info     ] DQN_20230313190141: epoch=2 step=5000 epoch=2 metrics={'time_sample_batch': 0.00010874042510986328, 'time_algorithm_update': 0.0030711585998535154, 'loss': 3.558228413348843e-05, 'time_step': 0.003266118335723877, 'td_error': 0.9870644139670596, 'value_scale': 1.0012083776692806, 'environment': 9.3} step=5000
2023-03-13 19:02:00 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_5000.pt


Epoch 3/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:02:10 [info     ] DQN_20230313190141: epoch=3 step=7500 epoch=3 metrics={'time_sample_batch': 0.00011303062438964843, 'time_algorithm_update': 0.0031888752937316894, 'loss': 3.1675507657382695e-05, 'time_step': 0.0033857173919677732, 'td_error': 0.9833814717893776, 'value_scale': 0.9939531941963918, 'environment': 17.3} step=7500
2023-03-13 19:02:10 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_7500.pt


Epoch 4/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:02:20 [info     ] DQN_20230313190141: epoch=4 step=10000 epoch=4 metrics={'time_sample_batch': 0.00011527967453002929, 'time_algorithm_update': 0.0031881381034851073, 'loss': 0.007066348025825027, 'time_step': 0.00339133882522583, 'td_error': 0.9918641160631149, 'value_scale': 2.0060413379784223, 'environment': 12.4} step=10000
2023-03-13 19:02:20 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_10000.pt


Epoch 5/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:02:29 [info     ] DQN_20230313190141: epoch=5 step=12500 epoch=5 metrics={'time_sample_batch': 0.00010668191909790038, 'time_algorithm_update': 0.003076361846923828, 'loss': 0.006230760553551954, 'time_step': 0.003267509746551514, 'td_error': 0.9752552312302291, 'value_scale': 1.98381710454265, 'environment': 19.9} step=12500
2023-03-13 19:02:29 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_12500.pt


Epoch 6/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:02:38 [info     ] DQN_20230313190141: epoch=6 step=15000 epoch=6 metrics={'time_sample_batch': 0.00010726566314697266, 'time_algorithm_update': 0.0031132880210876465, 'loss': 0.006048588067226229, 'time_step': 0.00330434045791626, 'td_error': 0.9825717391032092, 'value_scale': 1.9832849040915217, 'environment': 17.8} step=15000
2023-03-13 19:02:38 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_15000.pt


Epoch 7/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:02:48 [info     ] DQN_20230313190141: epoch=7 step=17500 epoch=7 metrics={'time_sample_batch': 0.00010572671890258789, 'time_algorithm_update': 0.003082563495635986, 'loss': 0.015362253805273212, 'time_step': 0.0032682649612426756, 'td_error': 1.0135379743300548, 'value_scale': 2.99799875456408, 'environment': 175.3} step=17500
2023-03-13 19:02:48 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_17500.pt


Epoch 8/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:02:57 [info     ] DQN_20230313190141: epoch=8 step=20000 epoch=8 metrics={'time_sample_batch': 0.00010021581649780273, 'time_algorithm_update': 0.0030054548263549804, 'loss': 0.01864413274436083, 'time_step': 0.0031805929183959963, 'td_error': 0.9966686411730681, 'value_scale': 2.9763200114003054, 'environment': 114.7} step=20000
2023-03-13 19:02:57 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_20000.pt


Epoch 9/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:03:07 [info     ] DQN_20230313190141: epoch=9 step=22500 epoch=9 metrics={'time_sample_batch': 0.00010303192138671875, 'time_algorithm_update': 0.0030933475494384767, 'loss': 0.01833996931213187, 'time_step': 0.003274539566040039, 'td_error': 0.9954841789499165, 'value_scale': 2.973896644438197, 'environment': 153.2} step=22500
2023-03-13 19:03:07 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_22500.pt


Epoch 10/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:03:17 [info     ] DQN_20230313190141: epoch=10 step=25000 epoch=10 metrics={'time_sample_batch': 0.00010579662322998047, 'time_algorithm_update': 0.0032065747261047365, 'loss': 0.024241683905909302, 'time_step': 0.0033939443588256834, 'td_error': 1.0467469998968637, 'value_scale': 3.9858706374229054, 'environment': 199.9} step=25000
2023-03-13 19:03:17 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_25000.pt


[(1,
  {'time_sample_batch': 0.00012184162139892578,
   'time_algorithm_update': 0.003392033576965332,
   'loss': 0.012645094241556217,
   'time_step': 0.00361170015335083,
   'td_error': 0.9881827529171064,
   'value_scale': 1.0015601515105368,
   'environment': 9.4}),
 (2,
  {'time_sample_batch': 0.00010874042510986328,
   'time_algorithm_update': 0.0030711585998535154,
   'loss': 3.558228413348843e-05,
   'time_step': 0.003266118335723877,
   'td_error': 0.9870644139670596,
   'value_scale': 1.0012083776692806,
   'environment': 9.3}),
 (3,
  {'time_sample_batch': 0.00011303062438964843,
   'time_algorithm_update': 0.0031888752937316894,
   'loss': 3.1675507657382695e-05,
   'time_step': 0.0033857173919677732,
   'td_error': 0.9833814717893776,
   'value_scale': 0.9939531941963918,
   'environment': 17.3}),
 (4,
  {'time_sample_batch': 0.00011527967453002929,
   'time_algorithm_update': 0.0031881381034851073,
   'loss': 0.007066348025825027,
   'time_step': 0.00339133882522583,
   '

### Metrics

**td_error** - larger td error means that the estimated state values are too far from the correct state values
The TD error is the difference between the correct reward based value and our estimate of the reward based value. The policy is the actor and the value function is the critic, with the ‘criticism’ being the TD error. If the TD error is positive the value of the action was greater than expected, suggesting the chosen action should be taken more often. If the TD error was negative the action had a lower value than expected, and so will be done less often in future states which are similar.

**discounted sum of average** - This is the discounted sum of advantage of the _**selected actions**_. This should be small to indicate that the policy selects actions with larger estimated values. This is the difference between the Q values of the dataset actions or the actions selected in each state and the max Q value according to the policy.  https://github.com/takuseno/d3rlpy/issues/99



In [51]:
observation = env.reset()

print(observation)

# return actions based on the greedy-policy
action = dqn.predict([observation])[0]

# estimate action-values
value = dqn.predict_value([observation], [action])

In [21]:
# This is the most useful metric for evaluating the progress of learning
def evaluate_mean_rewards(dqn):
    evaluate_scorer = evaluate_on_environment(env, n_trials=50)
    mean_episode_return = evaluate_scorer(dqn)
    return(mean_episode_return)

In [22]:
evaluate_mean_rewards(dqn)

NameError: name 'dqn' is not defined

In [54]:
value

In [49]:
dqn.predict([observation])

array([0])

### Generate offline data

In [145]:
import d3rlpy

# setup algorithm
random_policy = d3rlpy.algos.DiscreteRandomPolicy()

# prepare experience replay buffer
buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=500000, env=env)

# start data collection
random_policy.collect(env, buffer, n_steps=500000)

# export as MDPDataset
dataset = buffer.to_mdp_dataset()

# save MDPDataset
dataset.dump("random_policy_dataset_500000.h5")

2023-03-13 19:04:24 [debug    ] Building model...
2023-03-13 19:04:24 [debug    ] Model has been built.


  0%|          | 0/500000 [00:00<?, ?it/s]

In [33]:
ls

Intro_d3rlpy.ipynb                 random_policy_corridor_dataset.h5
d3rlpy_data/                       random_policy_dataset.h5
d3rlpy_logs/                       random_policy_dataset_80000.h5
poetry.lock                        trained_policy_dataset.h5
pyproject.toml                     trained_policy_dataset_80000.h5


In [146]:
# prepare experience replay buffer
buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=30000, env=env)

# start data collection
dqn.collect(env, buffer, n_steps=30000)

# export as MDPDataset
dataset = buffer.to_mdp_dataset()

# save MDPDataset
dataset.dump("trained_policy_dataset_30000.h5")

2023-03-13 19:04:58 [warning  ] Skip building models since they're already built.


  0%|          | 0/30000 [00:00<?, ?it/s]

import numpy as np

buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=80000, env=env)
dataset2 = buffer.to_mdp_dataset()
dataset2.load("random_policy_dataset_80000.h5")

buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=80000, env=env)
dataset3 = buffer.to_mdp_dataset()
dataset3.load("trained_policy_dataset_80000.h5")

In [147]:
dataset_trained = dataset.load("trained_policy_dataset_30000.h5")
dataset_random = dataset.load("random_policy_dataset_500000.h5")

In [148]:
train_episodes, test_episodes = train_test_split(dataset_trained, test_size=0.2)

In [149]:
dataset_trained.compute_stats()

{'return': {'mean': 197.7616,
  'std': 12.035108,
  'min': 51.0,
  'max': 199.0,
  'histogram': (array([  1,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
            0,   0,   0,   0,   0,   1, 149]),
   array([ 51. ,  58.4,  65.8,  73.2,  80.6,  88. ,  95.4, 102.8, 110.2,
          117.6, 125. , 132.4, 139.8, 147.2, 154.6, 162. , 169.4, 176.8,
          184.2, 191.6, 199. ], dtype=float32))},
 'reward': {'mean': 1.0,
  'std': 0.0,
  'min': 1.0,
  'max': 1.0,
  'histogram': (array([    0,     0,     0,     0,     0,     0,     0,     0,     0,
              0, 29862,     0,     0,     0,     0,     0,     0,     0,
              0,     0]),
   array([0.5 , 0.55, 0.6 , 0.65, 0.7 , 0.75, 0.8 , 0.85, 0.9 , 0.95, 1.  ,
          1.05, 1.1 , 1.15, 1.2 , 1.25, 1.3 , 1.35, 1.4 , 1.45, 1.5 ],
         dtype=float32))},
 'action': {'histogram': [[15465, 14397], array([0, 1])]},
 'observation': {'mean': array([-0.5493725 , -0.44689432, -0.0343954 , -0.02737486], dtype=float32),
  

In [150]:
dataset_random.compute_stats()

{'return': {'mean': 22.354317,
  'std': 11.985191,
  'min': 8.0,
  'max': 119.0,
  'histogram': (array([4730, 7077, 3742, 2736, 1372, 1043,  566,  433,  233,  155,  106,
            61,   38,   34,   17,   10,    4,    3,    4,    3]),
   array([  8.  ,  13.55,  19.1 ,  24.65,  30.2 ,  35.75,  41.3 ,  46.85,
           52.4 ,  57.95,  63.5 ,  69.05,  74.6 ,  80.15,  85.7 ,  91.25,
           96.8 , 102.35, 107.9 , 113.45, 119.  ], dtype=float32))},
 'reward': {'mean': 1.0,
  'std': 0.0,
  'min': 1.0,
  'max': 1.0,
  'histogram': (array([     0,      0,      0,      0,      0,      0,      0,      0,
               0,      0, 499999,      0,      0,      0,      0,      0,
               0,      0,      0,      0]),
   array([0.5 , 0.55, 0.6 , 0.65, 0.7 , 0.75, 0.8 , 0.85, 0.9 , 0.95, 1.  ,
          1.05, 1.1 , 1.15, 1.2 , 1.25, 1.3 , 1.35, 1.4 , 1.45, 1.5 ],
         dtype=float32))},
 'action': {'histogram': [[249843, 250156], array([0, 1])]},
 'observation': {'mean': array([2.318179

In [101]:
dataset_trained.actions

array([0, 0, 0, ..., 1, 0, 0], dtype=int32)

In [76]:
dataset_random.actions

array([1, 1, 1, ..., 0, 0, 0], dtype=int32)

In [57]:
dataset_trained.rewards

array([1., 1., 1., ..., 1., 1., 1.], dtype=float32)

In [152]:
len(dataset_trained.episodes)

151

In [151]:
len(dataset_random.episodes)

22367

In [79]:
dataset_trained.observations

array([[ 0.02962353, -0.0323091 , -0.03679555,  0.03139311],
       [ 0.02897735, -0.22688459, -0.03616769,  0.31224334],
       [ 0.02443966, -0.42147312, -0.02992282,  0.5933043 ],
       ...,
       [ 0.2347083 ,  0.5337355 , -0.13025276, -0.45787495],
       [ 0.24538301,  0.34067228, -0.13941026, -0.20891936],
       [ 0.25219646,  0.14779048, -0.14358865,  0.03674239]],
      dtype=float32)

### Train on offline data

In [153]:
from d3rlpy.algos import DQN
from d3rlpy.metrics.scorer import evaluate_on_environment
from d3rlpy.metrics.scorer import discounted_sum_of_advantage_scorer

# if you don't use GPU, set use_gpu=False instead.
dqn = DQN(use_gpu=False)

# initialize neural networks with the given observation shape and action size.
# this is not necessary when you directly call fit or fit_online method.
dqn.build_with_dataset(dataset_trained)

# calculate metrics with test dataset
td_error = td_error_scorer(dqn, test_episodes)

# set environment in scorer function, returns mean return per episode
evaluate_scorer = evaluate_on_environment(env)

# evaluate algorithm on the environment
rewards = evaluate_scorer(dqn)

# Use the output of learning, this is the mean reward and the most useful metric to measure learning
dqn.fit(train_episodes,
        eval_episodes=test_episodes,
        n_epochs=10,
        scorers={
            'td_error': td_error_scorer,
            'advantage': discounted_sum_of_advantage_scorer,
            'value_scale': average_value_estimation_scorer,
            'environment': evaluate_scorer
        })

2023-03-13 19:06:06 [debug    ] RoundIterator is selected.
2023-03-13 19:06:06 [info     ] Directory is created at d3rlpy_logs/DQN_20230313190606
2023-03-13 19:06:06 [warning  ] Skip building models since they're already built.
2023-03-13 19:06:06 [info     ] Parameters are saved to d3rlpy_logs/DQN_20230313190606/params.json params={'action_scaler': None, 'batch_size': 32, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 6.25e-05, 'n_critics': 1, 'n_frames': 1, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': None, 'target_update_interval': 8000, 'use_gpu': None, 'algorithm': 'DQN', 'observation_shape': (4,), 'action_size': 2}


Epoch 1/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:09 [info     ] DQN_20230313190606: epoch=1 step=742 epoch=1 metrics={'time_sample_batch': 0.00010505551597821423, 'time_algorithm_update': 0.002879885329069153, 'loss': 0.031179017385418388, 'time_step': 0.003060433421173507, 'td_error': 0.9846738294130843, 'advantage': -0.1300933121034453, 'value_scale': 1.0728607263069005, 'environment': 86.4} step=742
2023-03-13 19:06:09 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_742.pt


Epoch 2/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:12 [info     ] DQN_20230313190606: epoch=2 step=1484 epoch=2 metrics={'time_sample_batch': 0.00011077659792013246, 'time_algorithm_update': 0.0030716243137246515, 'loss': 3.344963113403049e-05, 'time_step': 0.003261560699689099, 'td_error': 0.985295856526353, 'advantage': -0.1917720939016169, 'value_scale': 1.0729103422340371, 'environment': 120.2} step=1484
2023-03-13 19:06:12 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_1484.pt


Epoch 3/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:15 [info     ] DQN_20230313190606: epoch=3 step=2226 epoch=3 metrics={'time_sample_batch': 0.00010746990574016726, 'time_algorithm_update': 0.003062619353240391, 'loss': 1.5665062441480843e-05, 'time_step': 0.0032469680045814206, 'td_error': 0.9840360184373476, 'advantage': -0.15674497508117402, 'value_scale': 1.0731224996386852, 'environment': 123.9} step=2226
2023-03-13 19:06:15 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_2226.pt


Epoch 4/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:18 [info     ] DQN_20230313190606: epoch=4 step=2968 epoch=4 metrics={'time_sample_batch': 0.00010607602461329047, 'time_algorithm_update': 0.003131949676657623, 'loss': 1.067789256195109e-05, 'time_step': 0.0033144237538874953, 'td_error': 0.9840877200697233, 'advantage': -0.15886232738107622, 'value_scale': 1.073022628412695, 'environment': 65.3} step=2968
2023-03-13 19:06:18 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_2968.pt


Epoch 5/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:21 [info     ] DQN_20230313190606: epoch=5 step=3710 epoch=5 metrics={'time_sample_batch': 0.00010663704730751059, 'time_algorithm_update': 0.0030635277216003913, 'loss': 8.527066199344161e-06, 'time_step': 0.0032455982223675256, 'td_error': 0.9836855233927463, 'advantage': -0.14836029094288894, 'value_scale': 1.073164146328116, 'environment': 81.3} step=3710
2023-03-13 19:06:21 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_3710.pt


Epoch 6/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:24 [info     ] DQN_20230313190606: epoch=6 step=4452 epoch=6 metrics={'time_sample_batch': 0.00010828509163663715, 'time_algorithm_update': 0.0030746199692677297, 'loss': 7.0954519173631595e-06, 'time_step': 0.003260059176750903, 'td_error': 0.9831215379138788, 'advantage': -0.12544745650668326, 'value_scale': 1.0730089680529256, 'environment': 99.3} step=4452
2023-03-13 19:06:24 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_4452.pt


Epoch 7/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:27 [info     ] DQN_20230313190606: epoch=7 step=5194 epoch=7 metrics={'time_sample_batch': 0.00010595553004838065, 'time_algorithm_update': 0.0029992861246484308, 'loss': 7.0120120548852535e-06, 'time_step': 0.0031801108722738177, 'td_error': 0.982881306147701, 'advantage': -0.11057738133555271, 'value_scale': 1.0711203961692182, 'environment': 152.6} step=5194
2023-03-13 19:06:27 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_5194.pt


Epoch 8/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:30 [info     ] DQN_20230313190606: epoch=8 step=5936 epoch=8 metrics={'time_sample_batch': 0.00011502250506871473, 'time_algorithm_update': 0.003311586187213579, 'loss': 7.016652208064585e-06, 'time_step': 0.003509262180071314, 'td_error': 0.9860360994536276, 'advantage': -0.22346773301398976, 'value_scale': 1.0732343780454097, 'environment': 149.1} step=5936
2023-03-13 19:06:30 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_5936.pt


Epoch 9/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:34 [info     ] DQN_20230313190606: epoch=9 step=6678 epoch=9 metrics={'time_sample_batch': 0.0001122112865396587, 'time_algorithm_update': 0.0031580108837939978, 'loss': 7.5880011481197564e-06, 'time_step': 0.003349788748029107, 'td_error': 0.9845790799986306, 'advantage': -0.17504556564075838, 'value_scale': 1.0736828844250994, 'environment': 169.0} step=6678
2023-03-13 19:06:34 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_6678.pt


Epoch 10/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:37 [info     ] DQN_20230313190606: epoch=10 step=7420 epoch=10 metrics={'time_sample_batch': 0.00011082286783305782, 'time_algorithm_update': 0.0030986314835252785, 'loss': 7.565187162005304e-06, 'time_step': 0.0032873475005363198, 'td_error': 0.9839800890170002, 'advantage': -0.1490605409202769, 'value_scale': 1.0740309759892235, 'environment': 189.7} step=7420
2023-03-13 19:06:37 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_7420.pt


[(1,
  {'time_sample_batch': 0.00010505551597821423,
   'time_algorithm_update': 0.002879885329069153,
   'loss': 0.031179017385418388,
   'time_step': 0.003060433421173507,
   'td_error': 0.9846738294130843,
   'advantage': -0.1300933121034453,
   'value_scale': 1.0728607263069005,
   'environment': 86.4}),
 (2,
  {'time_sample_batch': 0.00011077659792013246,
   'time_algorithm_update': 0.0030716243137246515,
   'loss': 3.344963113403049e-05,
   'time_step': 0.003261560699689099,
   'td_error': 0.985295856526353,
   'advantage': -0.1917720939016169,
   'value_scale': 1.0729103422340371,
   'environment': 120.2}),
 (3,
  {'time_sample_batch': 0.00010746990574016726,
   'time_algorithm_update': 0.003062619353240391,
   'loss': 1.5665062441480843e-05,
   'time_step': 0.0032469680045814206,
   'td_error': 0.9840360184373476,
   'advantage': -0.15674497508117402,
   'value_scale': 1.0731224996386852,
   'environment': 123.9}),
 (4,
  {'time_sample_batch': 0.00010607602461329047,
   'time_a

In [154]:
evaluate_mean_rewards(dqn)

118.18

In [155]:
train_episodes, test_episodes = train_test_split(dataset_random, test_size=0.2)

In [156]:
from d3rlpy.algos import DQN
from d3rlpy.metrics.scorer import evaluate_on_environment

# if you don't use GPU, set use_gpu=False instead.
dqn = DQN(use_gpu=False)

# initialize neural networks with the given observation shape and action size.
# this is not necessary when you directly call fit or fit_online method.
dqn.build_with_dataset(dataset_random)

# calculate metrics with test dataset
td_error = td_error_scorer(dqn, test_episodes)

# set environment in scorer function, returns mean return per episode
evaluate_scorer = evaluate_on_environment(env)

# evaluate algorithm on the environment
rewards = evaluate_scorer(dqn)

# Use the output of learning, this is the mean reward and the most useful metric to measure learning
dqn.fit(train_episodes,
        eval_episodes=test_episodes,
        n_epochs=10,
        scorers={
            'td_error': td_error_scorer,
            'value_scale': average_value_estimation_scorer,
            'environment': evaluate_scorer
        })

2023-03-13 19:07:50 [debug    ] RoundIterator is selected.
2023-03-13 19:07:50 [info     ] Directory is created at d3rlpy_logs/DQN_20230313190750
2023-03-13 19:07:50 [warning  ] Skip building models since they're already built.
2023-03-13 19:07:50 [info     ] Parameters are saved to d3rlpy_logs/DQN_20230313190750/params.json params={'action_scaler': None, 'batch_size': 32, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 6.25e-05, 'n_critics': 1, 'n_frames': 1, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': None, 'target_update_interval': 8000, 'use_gpu': None, 'algorithm': 'DQN', 'observation_shape': (4,), 'action_size': 2}


Epoch 1/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:08:41 [info     ] DQN_20230313190750: epoch=1 step=12473 epoch=1 metrics={'time_sample_batch': 0.00010740599291637906, 'time_algorithm_update': 0.0029705155300059948, 'loss': 0.007300398724316957, 'time_step': 0.003226995257877132, 'td_error': 0.9154510430734613, 'value_scale': 1.989812987434266, 'environment': 13.1} step=12473
2023-03-13 19:08:41 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_12473.pt


Epoch 2/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:09:30 [info     ] DQN_20230313190750: epoch=2 step=24946 epoch=2 metrics={'time_sample_batch': 9.262298969949896e-05, 'time_algorithm_update': 0.002876004726972957, 'loss': 0.019224161197751933, 'time_step': 0.0030891139496015984, 'td_error': 0.8774920830094897, 'value_scale': 3.8550574445173433, 'environment': 165.1} step=24946
2023-03-13 19:09:30 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_24946.pt


Epoch 3/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:10:22 [info     ] DQN_20230313190750: epoch=3 step=37419 epoch=3 metrics={'time_sample_batch': 9.637766657400884e-05, 'time_algorithm_update': 0.002961667989141947, 'loss': 0.025945931547490222, 'time_step': 0.0031868134792734336, 'td_error': 0.8805363140884244, 'value_scale': 4.753485571590386, 'environment': 129.4} step=37419
2023-03-13 19:10:22 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_37419.pt


Epoch 4/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:11:14 [info     ] DQN_20230313190750: epoch=4 step=49892 epoch=4 metrics={'time_sample_batch': 0.00010250079778033443, 'time_algorithm_update': 0.002999999839894651, 'loss': 0.029049355781752006, 'time_step': 0.0032261638990010904, 'td_error': 0.8977108749277593, 'value_scale': 6.4617853311665225, 'environment': 198.4} step=49892
2023-03-13 19:11:14 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_49892.pt


Epoch 5/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:12:07 [info     ] DQN_20230313190750: epoch=5 step=62365 epoch=5 metrics={'time_sample_batch': 0.00010686229227788273, 'time_algorithm_update': 0.0030226505797627855, 'loss': 0.033463577602898954, 'time_step': 0.003262710104888571, 'td_error': 0.9232798758382001, 'value_scale': 7.402156132463645, 'environment': 189.9} step=62365
2023-03-13 19:12:07 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_62365.pt


Epoch 6/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:13:02 [info     ] DQN_20230313190750: epoch=6 step=74838 epoch=6 metrics={'time_sample_batch': 0.00010990138846392583, 'time_algorithm_update': 0.0032613820294891166, 'loss': 0.04279645937602827, 'time_step': 0.003503515450275672, 'td_error': 1.022839158552869, 'value_scale': 9.176658419550185, 'environment': 148.8} step=74838
2023-03-13 19:13:02 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_74838.pt


Epoch 7/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:13:55 [info     ] DQN_20230313190750: epoch=7 step=87311 epoch=7 metrics={'time_sample_batch': 0.00010684241291267262, 'time_algorithm_update': 0.0031818968726171674, 'loss': 0.05209862822889459, 'time_step': 0.0034250694489905325, 'td_error': 1.0755130054085789, 'value_scale': 9.98658522042752, 'environment': 115.0} step=87311
2023-03-13 19:13:55 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_87311.pt


Epoch 8/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:14:47 [info     ] DQN_20230313190750: epoch=8 step=99784 epoch=8 metrics={'time_sample_batch': 0.00010092547277607915, 'time_algorithm_update': 0.0030981286109368305, 'loss': 0.06506972424293393, 'time_step': 0.0033279058211293034, 'td_error': 1.195821588186276, 'value_scale': 11.665359968979724, 'environment': 122.1} step=99784
2023-03-13 19:14:47 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_99784.pt


Epoch 9/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:15:40 [info     ] DQN_20230313190750: epoch=9 step=112257 epoch=9 metrics={'time_sample_batch': 0.00010414751646637793, 'time_algorithm_update': 0.0031380276820295117, 'loss': 0.07419467414129627, 'time_step': 0.0033796215691983553, 'td_error': 1.3563738346002665, 'value_scale': 13.410262844828551, 'environment': 112.3} step=112257
2023-03-13 19:15:40 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_112257.pt


Epoch 10/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:16:33 [info     ] DQN_20230313190750: epoch=10 step=124730 epoch=10 metrics={'time_sample_batch': 0.0001025821884890504, 'time_algorithm_update': 0.003151704972015677, 'loss': 0.08543142601732101, 'time_step': 0.0033856501969608357, 'td_error': 1.3679511803102369, 'value_scale': 14.214068009820659, 'environment': 113.1} step=124730
2023-03-13 19:16:33 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_124730.pt


[(1,
  {'time_sample_batch': 0.00010740599291637906,
   'time_algorithm_update': 0.0029705155300059948,
   'loss': 0.007300398724316957,
   'time_step': 0.003226995257877132,
   'td_error': 0.9154510430734613,
   'value_scale': 1.989812987434266,
   'environment': 13.1}),
 (2,
  {'time_sample_batch': 9.262298969949896e-05,
   'time_algorithm_update': 0.002876004726972957,
   'loss': 0.019224161197751933,
   'time_step': 0.0030891139496015984,
   'td_error': 0.8774920830094897,
   'value_scale': 3.8550574445173433,
   'environment': 165.1}),
 (3,
  {'time_sample_batch': 9.637766657400884e-05,
   'time_algorithm_update': 0.002961667989141947,
   'loss': 0.025945931547490222,
   'time_step': 0.0031868134792734336,
   'td_error': 0.8805363140884244,
   'value_scale': 4.753485571590386,
   'environment': 129.4}),
 (4,
  {'time_sample_batch': 0.00010250079778033443,
   'time_algorithm_update': 0.002999999839894651,
   'loss': 0.029049355781752006,
   'time_step': 0.0032261638990010904,
   't

In [157]:
evaluate_mean_rewards(dqn)

112.54

### Online learning

In [19]:
from gym.wrappers import TimeLimit
from d3rlpy.algos import DQN
from d3rlpy.online.buffers import ReplayBuffer
from d3rlpy.online.explorers import LinearDecayEpsilonGreedy

config={"corridor_length": 5}
env_corridor = SimpleCorridor_d3rlpy(config=config)

env = TimeLimit(env_corridor, max_episode_steps=1000)

#env = gym.make('CartPole-v0')
#eval_env = gym.make('CartPole-v0')

# setup algorithm
dqn = DQN(batch_size=32,
          learning_rate=2.5e-4,
          target_update_interval=100,
          use_gpu=False)

# setup replay buffer
buffer = ReplayBuffer(maxlen=100000, env=env)

# setup explorers
explorer = LinearDecayEpsilonGreedy(start_epsilon=1.0,
                                    end_epsilon=0.1,
                                    duration=10000)

# start training
dqn.fit_online(env,
               buffer,
               explorer=explorer, # you don't need this with probablistic policy algorithms
               eval_env=env,
               n_steps=3000, # the number of total steps to train.
               n_steps_per_epoch=1000,
               update_interval=10) # update parameters every 10 steps.

2023-03-13 23:21:48 [info     ] Directory is created at d3rlpy_logs/DQN_online_20230313232148
2023-03-13 23:21:48 [debug    ] Building model...
2023-03-13 23:21:48 [debug    ] Model has been built.
2023-03-13 23:21:48 [info     ] Parameters are saved to d3rlpy_logs/DQN_online_20230313232148/params.json params={'action_scaler': None, 'batch_size': 32, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 0.00025, 'n_critics': 1, 'n_frames': 1, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': None, 'target_update_interval': 100, 'use_gpu': None, 'algorithm': 'DQN', 'observation_shape': (1,), 'action_size': 5}


  0%|          | 0/3000 [00:00<?, ?it/s]

2023-03-13 23:21:49 [info     ] Model parameters are saved to d3rlpy_logs/DQN_online_20230313232148/model_1000.pt
2023-03-13 23:21:49 [info     ] DQN_online_20230313232148: epoch=1 step=1000 epoch=1 metrics={'time_inference': 0.0004203882217407227, 'time_environment_step': 8.6669921875e-06, 'time_step': 0.0007500715255737304, 'rollout_return': 1.426086956521738, 'time_sample_batch': 8.994525240868637e-05, 'time_algorithm_update': 0.002807388600614882, 'loss': 0.03939791176391324, 'evaluation': 2.3999999999999995} step=1000
2023-03-13 23:21:50 [info     ] Model parameters are saved to d3rlpy_logs/DQN_online_20230313232148/model_2000.pt
2023-03-13 23:21:50 [info     ] DQN_online_20230313232148: epoch=2 step=2000 epoch=2 metrics={'time_inference': 0.0004008944034576416, 'time_environment_step': 8.468389511108398e-06, 'rollout_return': 1.6809917355371897, 'time_step': 0.0007261524200439454, 'time_sample_batch': 9.036540985107421e-05, 'time_algorithm_update': 0.0026776432991027833, 'loss': 

In [20]:
from d3rlpy.metrics.scorer import evaluate_on_environment
evaluate_scorer = evaluate_on_environment(env, n_trials=50)
mean_episode_return = evaluate_scorer(dqn)
mean_episode_return

2.4000000000000004

### Summary

Samples from the random set do not result in good learning. You always want to generate samples using a trained policy. About 30000 samples (151 episodes) from a well trained policy gets better results than 500000 samples (22000 episodes) from a random policy. About 10x more samples and and more than 100x more samples in the random sampling. *More episodes are necessary for random samples because the mean length of these samples are small since they do not take ideal actions in this environment*. This behavior depends on the environment, the corridor example might have longer episodes in the random sampling case. The reward is really the only good metric to measure the performance of a learned algorithm.

Use the dataset summary to estimate the mean reward in both cases. The mean reward in trained is close to 200 while the random scenario is close to 110.

In [3]:
from d3rlpy.dataset import MDPDataset
from d3rlpy.algos import DiscreteCQL
from d3rlpy.datasets import get_cartpole
from sklearn.model_selection import train_test_split

dataset, env = get_cartpole()

# setup CQL algorithm (discrete version)
cql = DiscreteCQL(n_frames=4, scaler='pixel', use_gpu=False)

# split train and test episodes
dataset_trained = dataset.load("trained_policy_dataset_80000.h5")
train_episodes, test_episodes = train_test_split(dataset_trained, test_size=0.2)

In [11]:
from d3rlpy.metrics.scorer import discounted_sum_of_advantage_scorer
from d3rlpy.metrics.scorer import evaluate_on_environment
from d3rlpy.metrics.scorer import td_error_scorer
from d3rlpy.metrics.scorer import average_value_estimation_scorer

# start training
cql.fit(train_episodes,
        eval_episodes=test_episodes,
        n_epochs=20,
        scorers={
            'environment': evaluate_on_environment(env), # Cartpole environment
            'advantage': discounted_sum_of_advantage_scorer, # smaller is better
            'td_error': td_error_scorer, # smaller is better
            'value_scale': average_value_estimation_scorer # smaller is better
        })

2023-03-13 21:04:12 [debug    ] RoundIterator is selected.
2023-03-13 21:04:12 [info     ] Directory is created at d3rlpy_logs/DiscreteCQL_20230313210412
2023-03-13 21:04:12 [debug    ] Fitting scaler...              scaler=pixel
2023-03-13 21:04:12 [warning  ] Skip building models since they're already built.
2023-03-13 21:04:12 [info     ] Parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/params.json params={'action_scaler': None, 'alpha': 1.0, 'batch_size': 32, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 6.25e-05, 'n_critics': 1, 'n_frames': 4, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': {'type': 'pixel', 'params': {}}, 'target_upd

Epoch 1/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:04:22 [info     ] DiscreteCQL_20230313210412: epoch=1 step=2001 epoch=1 metrics={'time_sample_batch': 0.00010203898161545448, 'time_algorithm_update': 0.004655260732327623, 'loss': 0.6651579062680135, 'time_step': 0.0048200375196160465, 'environment': 91.8, 'advantage': -1.3254388459985096, 'td_error': 1.012708564675948, 'value_scale': 2.9697224840318843} step=2001
2023-03-13 21:04:22 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_2001.pt


Epoch 2/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:04:34 [info     ] DiscreteCQL_20230313210412: epoch=2 step=4002 epoch=2 metrics={'time_sample_batch': 0.00010622411534406137, 'time_algorithm_update': 0.004999962525031734, 'loss': 0.6580394860924869, 'time_step': 0.005170689410772519, 'environment': 48.6, 'advantage': -1.7185287662152486, 'td_error': 1.041537286662498, 'value_scale': 3.276492657850776} step=4002
2023-03-13 21:04:34 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_4002.pt


Epoch 3/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:04:44 [info     ] DiscreteCQL_20230313210412: epoch=3 step=6003 epoch=3 metrics={'time_sample_batch': 9.791294614533553e-05, 'time_algorithm_update': 0.004438420166557041, 'loss': 0.654462726934739, 'time_step': 0.004598321466669924, 'environment': 50.8, 'advantage': -2.2919969226923813, 'td_error': 1.070809157988891, 'value_scale': 4.001533606646148} step=6003
2023-03-13 21:04:44 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_6003.pt


Epoch 4/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:04:53 [info     ] DiscreteCQL_20230313210412: epoch=4 step=8004 epoch=4 metrics={'time_sample_batch': 9.198500953990778e-05, 'time_algorithm_update': 0.004164894719769632, 'loss': 0.6477935842309577, 'time_step': 0.00431484237186674, 'environment': 54.5, 'advantage': -2.5615374092175927, 'td_error': 1.0971446822715032, 'value_scale': 4.006084045708808} step=8004
2023-03-13 21:04:53 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_8004.pt


Epoch 5/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:05:03 [info     ] DiscreteCQL_20230313210412: epoch=5 step=10005 epoch=5 metrics={'time_sample_batch': 9.994004024141493e-05, 'time_algorithm_update': 0.004436529498884286, 'loss': 0.6438178669387612, 'time_step': 0.004598664736997956, 'environment': 61.6, 'advantage': -2.9490042279112116, 'td_error': 1.1298979617343161, 'value_scale': 4.018743174957206} step=10005
2023-03-13 21:05:03 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_10005.pt


Epoch 6/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:05:15 [info     ] DiscreteCQL_20230313210412: epoch=6 step=12006 epoch=6 metrics={'time_sample_batch': 0.00010785420259078225, 'time_algorithm_update': 0.005008210544881673, 'loss': 0.6437593708033564, 'time_step': 0.005184543901297642, 'environment': 160.1, 'advantage': -3.4741918671047007, 'td_error': 1.1553515195846558, 'value_scale': 4.449090934020331} step=12006
2023-03-13 21:05:15 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_12006.pt


Epoch 7/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:05:26 [info     ] DiscreteCQL_20230313210412: epoch=7 step=14007 epoch=7 metrics={'time_sample_batch': 0.00010124794665960477, 'time_algorithm_update': 0.004560264809497412, 'loss': 0.6558089409989754, 'time_step': 0.0047253140921833395, 'environment': 74.9, 'advantage': -3.3955624228666887, 'td_error': 1.1625632649999418, 'value_scale': 5.0369551028913655} step=14007
2023-03-13 21:05:26 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_14007.pt


Epoch 8/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:05:36 [info     ] DiscreteCQL_20230313210412: epoch=8 step=16008 epoch=8 metrics={'time_sample_batch': 0.00010204493910118915, 'time_algorithm_update': 0.004615444710467947, 'loss': 0.6547771691382378, 'time_step': 0.004783043797048314, 'environment': 79.1, 'advantage': -3.4426075962385885, 'td_error': 1.1583253843018095, 'value_scale': 5.062358342323656} step=16008
2023-03-13 21:05:36 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_16008.pt


Epoch 9/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:05:47 [info     ] DiscreteCQL_20230313210412: epoch=9 step=18009 epoch=9 metrics={'time_sample_batch': 9.496708859925506e-05, 'time_algorithm_update': 0.004242173080024929, 'loss': 0.6546474769972135, 'time_step': 0.004397149207531244, 'environment': 116.5, 'advantage': -3.6174741675999904, 'td_error': 1.1636514291760855, 'value_scale': 5.05985494430192} step=18009
2023-03-13 21:05:47 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_18009.pt


Epoch 10/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:05:57 [info     ] DiscreteCQL_20230313210412: epoch=10 step=20010 epoch=10 metrics={'time_sample_batch': 9.65741799510401e-05, 'time_algorithm_update': 0.004375766718882075, 'loss': 0.6570942673547335, 'time_step': 0.0045329779758863245, 'environment': 59.8, 'advantage': -3.494755354516521, 'td_error': 1.1580796078598092, 'value_scale': 5.706066122492553} step=20010
2023-03-13 21:05:57 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_20010.pt


Epoch 11/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:06:06 [info     ] DiscreteCQL_20230313210412: epoch=11 step=22011 epoch=11 metrics={'time_sample_batch': 9.489857751330634e-05, 'time_algorithm_update': 0.004219963930655217, 'loss': 0.6557715606653708, 'time_step': 0.004376385820799622, 'environment': 90.0, 'advantage': -3.516830491984502, 'td_error': 1.149193980818434, 'value_scale': 6.098671748288276} step=22011
2023-03-13 21:06:06 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_22011.pt


Epoch 12/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:06:16 [info     ] DiscreteCQL_20230313210412: epoch=12 step=24012 epoch=12 metrics={'time_sample_batch': 9.453957942293502e-05, 'time_algorithm_update': 0.004163691426800942, 'loss': 0.6550528301351491, 'time_step': 0.0043188941770646045, 'environment': 75.1, 'advantage': -3.4802442269634843, 'td_error': 1.152473978815058, 'value_scale': 6.090322522968782} step=24012
2023-03-13 21:06:16 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_24012.pt


Epoch 13/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:06:26 [info     ] DiscreteCQL_20230313210412: epoch=13 step=26013 epoch=13 metrics={'time_sample_batch': 9.184929801487196e-05, 'time_algorithm_update': 0.004145039015564544, 'loss': 0.6546901913060956, 'time_step': 0.004295159077298814, 'environment': 168.3, 'advantage': -3.9572124279550267, 'td_error': 1.16177161744492, 'value_scale': 6.0912839666303835} step=26013
2023-03-13 21:06:26 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_26013.pt


Epoch 14/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:06:36 [info     ] DiscreteCQL_20230313210412: epoch=14 step=28014 epoch=14 metrics={'time_sample_batch': 9.532680158791454e-05, 'time_algorithm_update': 0.00424109799214925, 'loss': 0.6572653990814175, 'time_step': 0.004397864820717693, 'environment': 131.6, 'advantage': -3.77304807076431, 'td_error': 1.1422299892552803, 'value_scale': 6.9785152498476615} step=28014
2023-03-13 21:06:36 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_28014.pt


Epoch 15/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:06:45 [info     ] DiscreteCQL_20230313210412: epoch=15 step=30015 epoch=15 metrics={'time_sample_batch': 9.432487163705745e-05, 'time_algorithm_update': 0.004216125641746083, 'loss': 0.6558399749719638, 'time_step': 0.004371163727104038, 'environment': 59.8, 'advantage': -4.402484999871009, 'td_error': 1.1669758997572262, 'value_scale': 7.141058743229809} step=30015
2023-03-13 21:06:45 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_30015.pt


Epoch 16/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:06:55 [info     ] DiscreteCQL_20230313210412: epoch=16 step=32016 epoch=16 metrics={'time_sample_batch': 9.242967627514368e-05, 'time_algorithm_update': 0.0041752900557777755, 'loss': 0.6555246020304686, 'time_step': 0.004328037011271891, 'environment': 180.5, 'advantage': -3.6602515807923086, 'td_error': 1.1357332365532302, 'value_scale': 7.134073428248599} step=32016
2023-03-13 21:06:55 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_32016.pt


Epoch 17/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:07:05 [info     ] DiscreteCQL_20230313210412: epoch=17 step=34017 epoch=17 metrics={'time_sample_batch': 9.44128041265012e-05, 'time_algorithm_update': 0.00420643912739065, 'loss': 0.6556290535852946, 'time_step': 0.004362281354173072, 'environment': 73.8, 'advantage': -3.4769954149638505, 'td_error': 1.1380020248988727, 'value_scale': 7.1318801294038545} step=34017
2023-03-13 21:07:05 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_34017.pt


Epoch 18/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:07:15 [info     ] DiscreteCQL_20230313210412: epoch=18 step=36018 epoch=18 metrics={'time_sample_batch': 9.2512723626285e-05, 'time_algorithm_update': 0.004176651222118433, 'loss': 0.6575932697258491, 'time_step': 0.004328266374472676, 'environment': 134.1, 'advantage': -3.6510411273301324, 'td_error': 1.121822730928422, 'value_scale': 8.15351369600926} step=36018
2023-03-13 21:07:15 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_36018.pt


Epoch 19/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:07:25 [info     ] DiscreteCQL_20230313210412: epoch=19 step=38019 epoch=19 metrics={'time_sample_batch': 9.792605261395181e-05, 'time_algorithm_update': 0.004505704010444424, 'loss': 0.6548346184838241, 'time_step': 0.00466829559196537, 'environment': 75.2, 'advantage': -3.476283405126625, 'td_error': 1.1227879632916373, 'value_scale': 8.129488967617567} step=38019
2023-03-13 21:07:25 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_38019.pt


Epoch 20/20:   0%|          | 0/2001 [00:00<?, ?it/s]

2023-03-13 21:07:35 [info     ] DiscreteCQL_20230313210412: epoch=20 step=40020 epoch=20 metrics={'time_sample_batch': 9.327992863919603e-05, 'time_algorithm_update': 0.004155320444445441, 'loss': 0.6543315184944454, 'time_step': 0.004309251867253324, 'environment': 61.1, 'advantage': -3.4743390531193312, 'td_error': 1.122848876645926, 'value_scale': 8.118136255179234} step=40020
2023-03-13 21:07:35 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230313210412/model_40020.pt


[(1,
  {'time_sample_batch': 0.00010203898161545448,
   'time_algorithm_update': 0.004655260732327623,
   'loss': 0.6651579062680135,
   'time_step': 0.0048200375196160465,
   'environment': 91.8,
   'advantage': -1.3254388459985096,
   'td_error': 1.012708564675948,
   'value_scale': 2.9697224840318843}),
 (2,
  {'time_sample_batch': 0.00010622411534406137,
   'time_algorithm_update': 0.004999962525031734,
   'loss': 0.6580394860924869,
   'time_step': 0.005170689410772519,
   'environment': 48.6,
   'advantage': -1.7185287662152486,
   'td_error': 1.041537286662498,
   'value_scale': 3.276492657850776}),
 (3,
  {'time_sample_batch': 9.791294614533553e-05,
   'time_algorithm_update': 0.004438420166557041,
   'loss': 0.654462726934739,
   'time_step': 0.004598321466669924,
   'environment': 50.8,
   'advantage': -2.2919969226923813,
   'td_error': 1.070809157988891,
   'value_scale': 4.001533606646148}),
 (4,
  {'time_sample_batch': 9.198500953990778e-05,
   'time_algorithm_update': 0.

In [13]:
evaluate_scorer = evaluate_on_environment(env, n_trials=50)
mean_episode_return = evaluate_scorer(cql)
mean_episode_return

79.22

### Custom environment

In [1]:
import gym
from gym.spaces import Discrete, Box
import numpy as np
import os
import random

class SimpleCorridor_d3rlpy(gym.Env):
    """Example of a custom env in which you have to walk down a corridor.
    Get a reward of -0.1 if you are not at the end, a random reward that is positive if you do.
    Move +1 if you move forward, -1 if you move backward. The total length is 5.
    We should want to reach the end in 5 steps in the perfectly trained world. 
    You can configure the length of the corridor via the env config."""

    def __init__(self, config):
        self.end_pos = config["corridor_length"]
        self.cur_pos = 0
        self.action_space = Discrete(5)
        self.observation_space = Box(0.0, self.end_pos, shape=(1,), dtype=np.float32)
        # Set the seed. This is only used for the final (reach goal) reward.
        self.reset()

    def reset(self, *, seed=None, options=None):
        random.seed(seed)
        self.cur_pos = 0
        return np.array([self.cur_pos])

    def step(self, action):
        assert action in [0, 1, 2, 3, 4], action
        # backward step
        if action == 0 and self.cur_pos > 0:
            self.cur_pos -= 1
            reward = -0.2
        # forward step
        elif action == 1:
            self.cur_pos += 1
            if(self.check_if_water()):
                reward = -0.5
            else:
                reward = 0.1
        # double speed
        elif action == 4:
            if(self.cur_pos <= 3):
                self.cur_pos += 2
            else:
                self.cur_pos += 1
            if(self.check_if_water()):
                reward = -0.5
            else:
                reward = 0.2
        # left or right
        else:
            self.cur_pos = self.cur_pos
            reward = -0.05
            
        done = truncated = self.cur_pos >= self.end_pos
        if(done):
            reward = 2
        # Produce a random reward when we reach the goal.
        return (
            np.array([self.cur_pos]),
            reward, # Setting to 2 instead of random reward has no real impact
            done,
            {},
        )
    
    def check_if_water(self):
        # This is water and will get a negative reward
        if(self.cur_pos == 3):
            return True
        else:
            return False

config={"corridor_length": 5}
env = SimpleCorridor_d3rlpy(config=config)

In [15]:
env

In [16]:
env.reset()

array([0])

In [5]:
action = env.action_space.sample()
obs, reward, done, info = env.step(action)

In [6]:
reward

-0.05

In [19]:
obs.astype

<function ndarray.astype>

In [41]:
import d3rlpy

# setup algorithm
random_policy = d3rlpy.algos.DiscreteRandomPolicy()

# prepare experience replay buffer
buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=100000, env=env)

# start data collection
random_policy.collect(corridor_env, buffer, n_steps=100000)

# export as MDPDataset
dataset = buffer.to_mdp_dataset()

# save MDPDataset
dataset.dump("random_policy_corridor_dataset.h5")

2023-03-13 21:33:12 [debug    ] Building model...
2023-03-13 21:33:12 [debug    ] Model has been built.


  0%|          | 0/100000 [00:00<?, ?it/s]

### Fit online with custom environment

In [113]:
# setup algorithm
dqn = d3rlpy.algos.DQN()

# prepare experience replay buffer
buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=100000, env=env)

# prepare exploration strategy if necessary
explorer = d3rlpy.online.explorers.ConstantEpsilonGreedy(0.3)

# start data collection
dqn.fit_online(env, buffer, n_steps=100000)

2023-03-10 14:47:32 [info     ] Directory is created at d3rlpy_logs/DQN_online_20230310144732
2023-03-10 14:47:32 [debug    ] Building model...
2023-03-10 14:47:32 [debug    ] Model has been built.
2023-03-10 14:47:32 [info     ] Parameters are saved to d3rlpy_logs/DQN_online_20230310144732/params.json params={'action_scaler': None, 'batch_size': 32, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 6.25e-05, 'n_critics': 1, 'n_frames': 1, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': None, 'target_update_interval': 8000, 'use_gpu': None, 'algorithm': 'DQN', 'observation_shape': (1,), 'action_size': 5}


  0%|          | 0/100000 [00:00<?, ?it/s]

2023-03-10 14:48:12 [info     ] Model parameters are saved to d3rlpy_logs/DQN_online_20230310144732/model_10000.pt
2023-03-10 14:48:12 [info     ] DQN_online_20230310144732: epoch=1 step=10000 epoch=1 metrics={'time_inference': 0.0005066035032272339, 'time_environment_step': 1.2087106704711915e-05, 'time_step': 0.003947368121147155, 'time_sample_batch': 0.0001312981424975613, 'time_algorithm_update': 0.00323518305445813, 'loss': 1.49070639037667e-06} step=10000
2023-03-10 14:48:50 [info     ] Model parameters are saved to d3rlpy_logs/DQN_online_20230310144732/model_20000.pt
2023-03-10 14:48:50 [info     ] DQN_online_20230310144732: epoch=2 step=20000 epoch=2 metrics={'time_inference': 0.0004898247003555298, 'time_environment_step': 1.1702489852905273e-05, 'time_sample_batch': 0.00012943384647369385, 'time_algorithm_update': 0.003080199909210205, 'loss': 7.911302969334136e-07, 'time_step': 0.0037821149587631228} step=20000
2023-03-10 14:49:30 [info     ] Model parameters are saved to d3

In [21]:
for i in range(5):
    obs = np.array([i])
    action = dqn.predict([obs])[0]
    print(action)

4
1
4
4
4


### Fit on offline data generated from custom environment

In [147]:
#initialize the algorithm
dqn = d3rlpy.algos.DQN()

dataset2 = d3rlpy.online.buffers.ReplayBuffer(maxlen=100000, env=env)
dataset2 = buffer.to_mdp_dataset()
dataset2.load("random_policy_corridor_dataset.h5")

train_episodes, test_episodes = train_test_split(dataset2, test_size=0.2)

In [ ]:
# start training
dqn.fit(train_episodes,
        eval_episodes=test_episodes,
        n_epochs=5,
        scorers={
            'environment': evaluate_on_environment(env), # Cartpole environment
            'advantage': discounted_sum_of_advantage_scorer, # smaller is better
            'td_error': td_error_scorer, # smaller is better
            'value_scale': average_value_estimation_scorer # smaller is better
        })

In [148]:
dqn.fit(
    train_episodes,
    n_epochs=10,
    #scorers=metrics,
    eval_episodes=test_episodes,
)